core/node/rpc/sync/client/message_router.go

```diff
@@ -0,0 +1,199 @@
+package client
+
+import (
+	"context"
+	"sync"
+
+	"github.com/towns-protocol/towns/core/node/rpc/sync/dynmsgbuf"
+
+	"github.com/puzpuzpuz/xsync/v4"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/towns-protocol/towns/core/node/logging"
+	. "github.com/towns-protocol/towns/core/node/protocol"
+	. "github.com/towns-protocol/towns/core/node/shared"
+)
+
+type subscription struct {
+	ctx      context.Context
+	cancel   context.CancelCauseFunc
+	syncID   string
+	messages *dynmsgbuf.DynamicBuffer[*SyncStreamsResponse]
+}
+
+func newSubscription(
+	ctx context.Context,
+	cancel context.CancelCauseFunc,
+	syncID string,
+) *subscription {
+	return &subscription{
+		ctx:      ctx,
+		cancel:   cancel,
+		syncID:   syncID,
+		messages: dynmsgbuf.NewDynamicBuffer[*SyncStreamsResponse](),
+	}
+}
+
+func (s *subscription) Close(err error) {
+	s.messages.Close()
+	s.cancel(err)
+}
+
+type messageRouter struct {
+	log                 *logging.Log
+	subscriptions       *xsync.Map[string, *subscription]
+	streamSubscriptions *xsync.Map[StreamId, *xsync.Map[string, struct{}]]
+	mu                  sync.Mutex // for shutdown
+	shutdown            bool
+}
+
+func newMessageRouter(ctx context.Context) *messageRouter {
+	return &messageRouter{
+		log:                 logging.FromCtx(ctx).Named("message_router"),
+		subscriptions:       xsync.NewMap[string, *subscription](),
+		streamSubscriptions: xsync.NewMap[StreamId, *xsync.Map[string, struct{}]](),
+	}
+}
+
+func (r *messageRouter) registerSubscription(syncID string, sub *subscription) {
+	r.subscriptions.Store(syncID, sub)
+}
+
+func (r *messageRouter) unregisterSubscription(syncID string) {
+	r.subscriptions.Delete(syncID)
+	r.streamSubscriptions.Range(func(streamID StreamId, syncIDs *xsync.Map[string, struct{}]) bool {
+		syncIDs.Delete(syncID)
+		if syncIDs.Size() == 0 {
+			r.streamSubscriptions.Delete(streamID)
+		}
+		return true
+	})
+}
+
+func (r *messageRouter) addStreamToSubscription(syncID string, streamID StreamId) {
+	syncIDs, _ := r.streamSubscriptions.LoadOrStore(streamID, xsync.NewMap[string, struct{}]())
+	syncIDs.Store(syncID, struct{}{})
+	r.streamSubscriptions.Store(streamID, syncIDs)
+}
+
+func (r *messageRouter) removeStreamFromSubscription(syncID string, streamID StreamId) {
+	if subscribers, ok := r.streamSubscriptions.Load(streamID); ok {
+		subscribers.Delete(syncID)
+		if subscribers.Size() == 0 {
+			r.streamSubscriptions.Delete(streamID)
```

Comment:
There is a race condition with `messageRouter#addStreamToSubscription` that can call `Store` after the size check and before the delete.

In general the usage of streamSubscriptions looks racy over the various functions unless messageRouter will be locked externally.